src/openforms/js/components/admin/form_design/data/complete-form.js

```diff
@@ -0,0 +1,276 @@
+  if (!hasErrors) {
+    // delete the steps marked for deletion
+    // TODO: error handling in case this fails - the situation before refactor
+    // was also a bit dire, the internal state was never cleaned up.
+    await Promise.all(stepsToDelete.map(async step => await apiDelete(step, csrftoken)));
+    newState = produce(newState, draft => {
+      draft.stepsToDelete = [];
+    });
+  }
+
+  return [newState, validationErrors];
+};
+
+/**
+ * Save the logic rules and price rules, report back any validation errors.
+ */
+const saveLogic = async (state, csrftoken) => {
+  // form logic
+  const {
+    form: {url: formUrl},
+    logicRules,
+    logicRulesToDelete,
+    priceRules,
+    priceRulesToDelete,
+  } = state;
+
+  const logicResults = await saveLogicRules(formUrl, csrftoken, logicRules, logicRulesToDelete);
+  const priceResults = await savePriceRules(formUrl, csrftoken, priceRules, priceRulesToDelete);
+
+  // update the state with updated references
+  const validationErrors = [];
+  const newState = produce(state, draft => {
+    for (const [index, result] of logicResults.entries()) {
+      if (result instanceof ValidationErrors) {
+        validationErrors.push(result);
+        continue;
+      }
+
+      const {url, uuid} = result;
+      draft.logicRules[index].uuid = uuid;
+      draft.logicRules[index].url = url;
+    }
+
+    for (const [index, result] of priceResults.entries()) {
+      if (result instanceof ValidationErrors) {
+        validationErrors.push(result);
+        continue;
+      }
+
+      const {url, uuid} = result;
+      draft.priceRules[index].uuid = uuid;
+      draft.priceRules[index].url = url;
+    }
+
+    draft.logicRulesToDelete = [];
+    draft.priceRulesToDelete = [];
+  });
+
+  return [newState, validationErrors];
+};
+
+/**
+ * Save the variables belonging to the form, user defined and derived from
+ * the form steps.
+ */
+const saveVariables = async (state, csrftoken) => {
+  const {
+    form: {url: formUrl},
+  } = state;
+
+  // resolve the variable.formDefinition URLs from the updated state after form steps
+  // and form definitions have been saved. This relies on the _generatedId for temporary
+  // steps/definitions.
+  const stepsByGeneratedId = Object.fromEntries(
+    state.formSteps.filter(step => !!step._generatedId).map(step => [step._generatedId, step])
+  );
+  let newState = produce(state, draft => {
+    for (const variable of draft.formVariables) {
+      variable.form = formUrl;
+      if (variable.formDefinition == null) continue;
+      // if the variable.formDefinition is not a URL, we have to resolve it against the
+      // temporary client-side ID. This also allows us to update the state with the
+      // actual resolved resource URLs.
+      try {
+        new URL(variable.formDefinition);
+      } catch {
+        const formStep = stepsByGeneratedId[variable.formDefinition];
+        variable.formDefinition = formStep.formDefinition;
```

Comment:
In the case where I add a user defined variable with nothing filled in (invalid) and then I save the form, I see an error in the console => formStep is undefined

Reply:
Fixed, was the cause of https://github.com/open-formulieren/open-forms/pull/1767#discussion_r919870341 as well